packages/flutter/lib/src/material/bottom_navigation.dart

```diff
@@ -0,0 +1,536 @@
+import 'dart:collection' show Queue;
+
+import 'package:flutter/widgets.dart';
+import 'package:vector_math/vector_math_64.dart' show Vector3;
+
+import 'colors.dart';
+import 'icon.dart';
+import 'icon_theme.dart';
+import 'icon_theme_data.dart';
+import 'ink_well.dart';
+import 'material.dart';
+import 'theme.dart';
+
+const double kBottomNavigationHeight = 60.0;
+const double _kActiveMaxWidth = 168.0;
+const double _kInactiveMaxWidth = 96.0;
+
+/// Closure used in [BottomNavigation] to handle action taps. The [animation]
+/// argument is useful in order to coordinate with the action's animation. The
+/// [index] argument specifies which action has been tapped.
+typedef void BottomNavigationCallback(Animation<double> animation, int index);
+
+/// The [BottomNavigation] type which affects its layout.
+///
+/// See also:
+///
+///  * [BottomNavigation]
+///  * <https://material.google.com/components/bottom-navigation.html#bottom-navigation-specs>
+enum BottomNavigationType {
+  /// Bottom navigation actions with fixed width.
+  fixed,
+  /// Bottom navigation actions with animated, shifting width.
+  shifting
+}
+
+/// An interactive action within [BottomNavigation] with an icon and label.
+///
+/// See also:
+///
+///  * [BottomNavigation]
+///  * <https://material.google.com/components/bottom-navigation.html>
+class BottomNavigationAction {
+  /// Creates an action that is used with [BottomNavigation].
+  ///
+  /// The arguments [icon] and [label] should not be null.
+  BottomNavigationAction({
+    this.icon,
+    this.label,
+    this.backgroundColor
+  }) {
+    assert(this.icon != null);
+    assert(this.label != null);
+  }
+
+  /// The icon of the action. Needs manual styling in the
+  /// [BottomNavigationType.fixed] case.
+  final Icon icon;
+
+  /// The label of the action. Needs manual styling in the
+  /// [BottomNavigationType.fixed] case.
+  final Widget label;
+
+  /// The radially growing background color of the action. It is only
+  /// appliccable in the [BottomNavigationType.shifting] case.
+  final Color backgroundColor;
+}
+
+/// A material widget for bottom navigation with ineractive actions.
+///
+/// See also:
+///
+///  * [BottomNavigationAction]
+///  * [Scaffold]
+///  * <https://material.google.com/components/bottom-navigation.html>
+class BottomNavigation extends StatefulWidget {
+  /// Creates a bottom navigation bar, typically used whithin a [Scaffold].
+  ///
+  /// The arguments [actions] and [type] should not be null.
+  BottomNavigation({
+    this.actions,
+    this.onAction,
+    this.type: BottomNavigationType.fixed
+  }) {
+    assert(this.actions != null);
+    assert(this.type != null);
+  }
+
+  /// The interactive actions layed out within the bottom navigation bar.
+  final List<BottomNavigationAction> actions;
+
+  /// The callback that is called when an action is tapped.
+  final BottomNavigationCallback onAction;
+
+  /// The type of the bottom navigation that effects its layout.
```

Comment:
Replace this with the description of what BottomNavigationType means from above.